internal/pkg/cli/app_deploy.go

```diff
@@ -348,46 +354,92 @@ func (o *appDeployOpts) getAppDeployTemplate() (string, error) {
 			GlobalOpts: o.GlobalOpts,
 		},
 
-		stackWriter:   buffer,
+		stackWriter:   appBuffer,
 		paramsWriter:  ioutil.Discard,
-		addonsWriter:  ioutil.Discard,
+		addonsWriter:  addonsBuffer,
 		initAddonsSvc: initPackageAddonsSvc,
 		store:         o.projectService,
 		describer:     o.appPackageCfClient,
 		ws:            o.workspaceService,
 	}
 
 	if err := appPackage.Execute(); err != nil {
-		return "", fmt.Errorf("package application: %w", err)
+		return nil, nil, fmt.Errorf("retrieve template: %w", err)
 	}
-	return buffer.String(), nil
+	return appBuffer, addonsBuffer, nil
 }
 
-func (o *appDeployOpts) applyAppDeployTemplate(template, stackName, changeSetName, cfExecutionRole string, tags map[string]string) error {
-	if err := o.appDeployCfClient.DeployApp(template, stackName, changeSetName, cfExecutionRole, tags); err != nil {
-		return fmt.Errorf("deploy application: %w", err)
+func (o *appDeployOpts) pushAddonsTemplateToS3Bucket(addonsTemplate *bytes.Buffer) error {
+	if !o.EnableAddons || addonsTemplate.String() == "" {
+		return nil
+	}
+	proj, err := o.projectService.GetProject(o.ProjectName())
+	if err != nil {
+		return fmt.Errorf("get project: %w", err)
 	}
+	resources, err := o.appPackageCfClient.GetProjectResourcesByRegion(proj, o.targetEnvironment.Region)
+	if err != nil {
+		return fmt.Errorf("get project resources: %w", err)
+	}
+
+	artifactRelativePath, err := o.s3Service.PutArtifact(resources.S3Bucket, fmt.Sprintf(archer.AddonsCfnTemplateNameFormat, o.AppName), addonsTemplate)
+	if err != nil {
+		return fmt.Errorf("put addons artifact to bucket %s: %w", resources.S3Bucket, err)
+	}
+	o.addonsArtifactPath = fmt.Sprintf(addonsURLPattern, resources.S3BucketRegionalDomainName, artifactRelativePath)
```

Comment:
What do you think of not storing this in the state? It seems like a hidden side-effect. Instead, we can have:
```go
// pushAddonsTemplateToS3Bucket uploads the template to the pipeline s3 bucket and returns the key of the object.
func (o *appDeployOpts) pushAddonsTemplateToS3Bucket(tpl bytes.Buffer) (string, error)

// deployAppStack creates a CFN stack given the application template. 
// If the addonsTplURL is not empty, it will also pass it as a parameter for the nested stack.
func (o *appDeployOpts) deployAppStack(tpl bytes.Buffer, addonsTplURL string) error
```